added support for SCA with Polyspace

mbeckersys opened the pull request:
This change adds support for the Polyspace Static Code Analysis tool on Zephyr. 
When analysis finishes, the results are printed on the console. Usage examples are provided, with an option to analyze only the user code.

mbeckersys commented:
Fixed commit message according to guidelines.

mbeckersys commented:
Fixed coding guidelines

mbeckersys commented:
Sorry for the noise, it took me some time to understand how to run compliance checks before push locally, and to see what needs to be fixed. All done now.

pdgendt commented on cmake/sca/polyspace/sca.cmake:
```diff
@@ -0,0 +1,145 @@
+    ${POLYSPACE_OPTIONS_ZEPHYR}
+    ${POLYSPACE_OPTIONS_FILE}
+    ${POLYSPACE_OPTIONS}
+    #    || ${CMAKE_COMMAND} -E true # allow to continue processing results
+    #VERBATIM
```
Remove if not needed

pdgendt commented on cmake/sca/polyspace/sca.cmake:
```diff
@@ -0,0 +1,145 @@
+  USES_TERMINAL
+)
+
+# EOF
```
Comment not that useful

pdgendt commented on cmake/sca/polyspace/polyspace-print-console.py:
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/python3
+"""
+Example script to illustrate the usage of the Polyspace Python tools
+Copyright (C) 2020-2024 The MathWorks, Inc.
```
Missing Apache-2.0 SPDX comment

pdgendt commented:
Do not add fixup commits, squash into a single commit and force push.

kartben commented on cmake/sca/polyspace/README.md:
we don't really do markdown, plus, this would be better as proper user facing documentation showing up in https://docs.zephyrproject.org/latest/develop/sca/index.html#native-sca-tool-support :)
Please consider moving/transforming your README.rst to doc/develop/sca/polyspace.rst and use the same document structure as the files there (ex. codechecker.rst)

mbeckersys replied:
Done

kartben reviewed (changes requested):
Nice, thanks! Please see my comment regarding the documentation

pdgendt commented on doc/develop/sca/polyspace.rst:
You also need to add an entry to `doc/develop/sca/index.rst`.

pdgendt commented on cmake/sca/polyspace/sca.cmake:
```diff
@@ -0,0 +1,142 @@
+find_program(POLYSPACE_RESULTS_EXE NAMES polyspace-results-export REQUIRED)
+
+
+# Get Polyspace speciic variables
```
```suggestion
# Get Polyspace specific variables
```

pdgendt commented on cmake/sca/polyspace/polyspace-print-console.py:
```diff
@@ -0,0 +1,88 @@
+    if args.details:
+        for f in findings:
+            print(
+                "{}:{}:{}: {} {}".format(
+                    f.get('File', 'unknown file'),
+                    f.get('Line', '?'),
+                    f.get('Col', '?'),
+                    f.get('Family', ''),
+                    f.get('Check', 'Unknown'),
+                )
```
Prefer to use f-strings instead of `str.format`.

mbeckersys replied:
I find it much less readable, and it hasn't been flagged by ruff. But I'll make the change for consistency

pdgendt commented on cmake/sca/polyspace/polyspace-print-console.py:
```diff
@@ -0,0 +1,88 @@
+
+    findings = []
+    header = []
+    with open(filename, encoding="latin-1") as fp:
```
Is the result file always in this encoding? Not an issue perse, just odd as utf-8 is considered the default in many areas.

mbeckersys replied:
We indeed need latin-1 here.

pdgendt commented on cmake/sca/polyspace/polyspace-print-console.py:
```diff
@@ -0,0 +1,85 @@
+        for f in findings:
+            print(
+                f"{f.get('File', 'unknown file')}:"
+                + "{f.get('Line', '?')}:"
+                + "{f.get('Col', '?')}: "
+                + "{f.get('Family', '')} {f.get('Check', 'Defect')}"
```
```suggestion
                + f"{f.get('Line', '?')}:"
                + f"{f.get('Col', '?')}: "
                + f"{f.get('Family', '')} {f.get('Check', 'Defect')}"
```

mbecker-netizen committed:
sca: added support for Polyspace tool

Add the cmake files for running static code analysis with the Polyspace
tools in the west build. The analysis leverages the compilation database.
Options for the analysis are documented in doc/develop/sca/polyspace.rst.

Analysis results are printed as command line output and provided as CSV.

Manually tested on v4.0.0 with various sample applications.

Signed-off-by: Martin Becker <mbecker@mathworks.com>

pdgendt reviewed (approved):
Thanks!
